ttnn/cpp/ttnn/operations/moreh/moreh_fold/device/kernels/reader_fold_rm.cpp

```diff
@@ -0,0 +1,109 @@
+    constexpr uint32_t input_cb_id = get_compile_time_arg_val(1);
+    constexpr uint32_t output_cb_id = get_compile_time_arg_val(2);
+    constexpr uint32_t onetile = 1;
+
+    uint32_t P = kernel_size_h * kernel_size_w;
+    uint32_t l = LH * LW;
+
+    const InterleavedAddrGen<input_is_dram> s0 = {.bank_base_address = input_addr, .page_size = input_cb_page_size};
+
+    for (uint32_t row_id = start_id; row_id < start_id + num_units_per_core; row_id++) {
+        cb_reserve_back(output_cb_id, onetile);
+        for (uint32_t elem_id = 0; elem_id < W; elem_id++) {
+            uint32_t gid = row_id * W + elem_id;
+            uint32_t nch = gid / W;
+            uint32_t w = gid % W;
+            uint32_t nc = nch / H;
+            uint32_t h = nch % H;
+            uint32_t n = nc / C;
+            uint32_t c = nc % C;
+
+            float sum = 0.0;
+            for (uint32_t ph = 0; ph < kernel_size_h; ++ph) {
+                for (uint32_t pw = 0; pw < kernel_size_w; ++pw) {
+                    uint32_t lhsh = h - ph * dilation_h + padding_h;
+                    uint32_t lwsw = w - pw * dilation_w + padding_w;
+                    if (lhsh % stride_h != 0)
+                        continue;
+                    if (lwsw % stride_w != 0)
+                        continue;
+                    uint32_t lh = lhsh / stride_h;
+                    uint32_t lw = lwsw / stride_w;
+                    if (lh < 0 || LH <= lh)
+                        continue;
+                    if (lw < 0 || LW <= lw)
+                        continue;
+
+                    // Input size is {N, C * kernel_size_h * kernel_size_w, LH * LW}
+                    uint32_t input_row_id = n * C * P + (c * P + ph * kernel_size_w + pw);
+                    // Read entire row into input_cb
+                    cb_reserve_back(input_cb_id, onetile);
+                    uint32_t l1_write_addr = get_write_ptr(input_cb_id);
+                    uint64_t src_noc_addr = get_noc_addr(input_row_id, s0);
+                    noc_async_read(src_noc_addr, l1_write_addr, input_cb_page_size);
+                    noc_async_read_barrier();
+                    cb_push_back(input_cb_id, onetile);
+
+                    cb_wait_front(input_cb_id, onetile);
+                    // TODO: Duong double check the DTYPE_BFLOAT16 conversion to float
+                    // This cause precision problems when the input and output is large (a lot of calculations)
+#ifdef DTYPE_BFLOAT16
+                    uint16_t *input_cb_ptr_uint16 = reinterpret_cast<uint16_t *>(get_read_ptr(input_cb_id));
+                    uint16_t bfloat16_value = input_cb_ptr_uint16[lh * LW + lw];
+                    uint32_t float_value_as_int = static_cast<uint32_t>(bfloat16_value) << 16;
+                    auto tmp = reinterpret_cast<float *>(&float_value_as_int);
+                    float value_as_float = *tmp;
+                    sum += value_as_float;
```

Comment:
It seems there are two different implementations of `bfloat16_to_float` currently in TT.
Could you check if the results change when using these `bfloat16 to _float` implementations?

1.
```C++
static inline float bfloat16_to_float(uint16_t bfloat_val) {
    uint32_t uint32_data = ((uint32_t)bfloat_val) << 16;
    float f;
    std::memcpy(&f, &uint32_data, sizeof(f));
    return f;
}
```

2.
```C++
float bfloat16_to_float32(uint16_t bfloat16_data) {
    uint32_t bits = static_cast<uint32_t>(bfloat16_data) << 16;

    // Extract the sign bit
    uint32_t sign = bits & 0x80000000;

    // Extract the exponent
    uint32_t exponent = bits & 0x7F800000;

    // Extract the mantissa
    uint32_t mantissa = bits & 0x007FFFFF;

    // Handle special cases
    if (exponent == 0 && mantissa == 0) {
        // Zero
        return sign ? -0.0f : 0.0f;
    } else if (exponent == 0x7F800000) {
        if (mantissa == 0) {
            // Infinity
            return sign ? -__builtin_huge_valf() : __builtin_huge_valf();
        } else {
            // NaN
            return __builtin_nanf("");
        }
    }

    // Assemble the float
    union {
        uint32_t u;
        float f;
    } ieee_float;

    ieee_float.u = sign | exponent | mantissa;
    return ieee_float.f;
}
```

Reply:
I use `float_to_bfloat16` to store the sum value to output_cb and I have tested both implementation of bf16 to float you mentioned above, and it still fails big tests. All smaller tests still passed.
```
static inline uint16_t float_to_bfloat16(float val) {
    union {
        float f;
        uint32_t u;
    } ret;
    ret.f = val;
    return uint16_t(ret.u >> 16);
}
```

Reply:
You mean that unit test of this implementation fails in some cases regardless of which typecasting function you use?

Reply:
Resolved in the latest commit: I have modified the pytest configuration to ensure that torch.rand generates primarily positive values. This adjustment prevents the cumulative sum of the output from approaching zero, allowing for more flexibility when comparing results using rtol and atol.